MoltenVK/MoltenVK/Commands/MVKCmdAccelerationStructure.mm

```diff
@@ -0,0 +1,200 @@
+
+VkResult MVKCmdCopyAccelerationStructure::setContent(MVKCommandBuffer*                  cmdBuff,
+                                                     VkAccelerationStructureKHR         srcAccelerationStructure,
+                                                     VkAccelerationStructureKHR         dstAccelerationStructure,
+                                                     VkCopyAccelerationStructureModeKHR copyMode) {
+    
+    MVKAccelerationStructure* mvkSrcAccStruct = (MVKAccelerationStructure*)srcAccelerationStructure;
+    MVKAccelerationStructure* mvkDstAccStruct = (MVKAccelerationStructure*)dstAccelerationStructure;
+    
+    _srcAccelerationStructure = mvkSrcAccStruct->getMTLAccelerationStructure();
+    _dstAccelerationStructure = mvkDstAccStruct->getMTLAccelerationStructure();
+    _copyMode = copyMode;
+    return VK_SUCCESS;
+}
+
+void MVKCmdCopyAccelerationStructure::encode(MVKCommandEncoder* cmdEncoder) {
+    id<MTLAccelerationStructureCommandEncoder> accStructEncoder = cmdEncoder->getMTLAccelerationStructureEncoder(kMVKCommandUseNone);
+    
+    if(_copyMode == VK_COPY_ACCELERATION_STRUCTURE_MODE_COMPACT_KHR)
+    {
+        [accStructEncoder
+         copyAndCompactAccelerationStructure:_srcAccelerationStructure
+         toAccelerationStructure:_dstAccelerationStructure];
+        
+        return;
+    }
+    
+    [accStructEncoder
+         copyAccelerationStructure:_srcAccelerationStructure
+         toAccelerationStructure:_dstAccelerationStructure];
+}
+
+#pragma mark -
+#pragma mark MVKCmdCopyAccelerationStructureToMemory
+
+VkResult MVKCmdCopyAccelerationStructureToMemory::setContent(MVKCommandBuffer*                  cmdBuff,
+                                                             VkAccelerationStructureKHR         srcAccelerationStructure,
+                                                             uint64_t                           dstAddress,
+                                                             VkCopyAccelerationStructureModeKHR copyMode) {
+    _dstAddress = dstAddress;
+    _copyMode = copyMode;
+    
+    MVKAccelerationStructure* mvkSrcAccStruct = (MVKAccelerationStructure*)srcAccelerationStructure;
+    _srcAccelerationStructure = mvkSrcAccStruct->getMTLAccelerationStructure();
+    
+    _dstBuffer = _mvkDevice->getBufferAtAddress(_dstAddress);
+    return VK_SUCCESS;
+}
+                                        
+void MVKCmdCopyAccelerationStructureToMemory::encode(MVKCommandEncoder* cmdEncoder) {
+    id<MTLAccelerationStructureCommandEncoder> accStructEncoder = cmdEncoder->getMTLAccelerationStructureEncoder(kMVKCommandUseNone);
+    _mvkDevice = cmdEncoder->getDevice();
+    
+    if(_copyMode != VK_COPY_ACCELERATION_STRUCTURE_MODE_SERIALIZE_KHR || !_dstBuffer->getDeviceMemory()->isMemoryHostAccessible()){
+        return;
+    }
+    
+    memcpy(_dstBuffer->getDeviceMemory()->getHostMemoryAddress(), (void*)_srcAccelerationStructure, sizeof(_srcAccelerationStructure));
```

Comment:
The copy is supposed to be performed on the device, especially since it is possible that one or both of the resources are in device-local (i.e. `MTLStorageModePrivate`) memory. The `-[MTLBlitCommandEncoder copyFromBuffer:sourceOffset:toBuffer:destinationOffset:size:]` method is what you want here.

Also, you'll need a method on the `MVKAccelerationStructure` to get its underlying `MVKBuffer`. And, you'll need to make sure that the `MTLAccelerationStructure` and `MTLBuffer` underlying both objects have their underlying GPU storage aliased to each other for this to work right. The only way I can think of to ensure that is to use a placement `MTLHeap`. Unfortunately, this means this functionality will now be limited to macOS Ventura or iOS 16, in which the `-[MTLHeap newAccelerationStructureWithDescriptor:offset:]` method you need for that was introduced.

Note that, although MoltenVK has support for using placement `MTLHeap`s, it's currently disabled due to some rendering artifacts with render targets in heaps. (I really need to check if that's been fixed.) But that doesn't matter, because you need to use an `MTLHeap` even in the dedicated allocation case.

Reply:
If I understand this correctly, you're proposing to implement Acceleration Structure copies using memory aliasing with a buffer and copying the buffer contents.

Why not just AccelerationStructureCommandEncoder::copy?

Reply:
This particular copy is for copying (i.e. serializing) acceleration structures to buffers.